### ISSUE TITLE
Expose API for Registering Post-Gradient-Computation Hook

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#84057]
### 🚀 The feature, motivation and pitch

To preface, this is not urgent or blocking.

Context: For FSDP to reduce-scatter gradients, it wants to register a hook that fires _after_ the gradient is computed. Currently, this is achieved by registering a hook on the `AccumulateGrad` object. This issue relates to how to access that `AccumulateGrad` object.

FSDP currently uses an `expand_as()` trick to bypass the fact that the `FlatParameter` (a leaf tensor) does not have a `grad_fn`. In the following code, `p_tmp` will have a `grad_fn` of `ExpandBackward0`, which then gives access to the `AccumulateGrad` object via `next_functions[0][0]`.
https://github.com/pytorch/pytorch/blob/00cb184512f3a636d87793f46d3f9c7fea406b25/torch/distributed/fsdp/fully_sharded_data_parallel.py#L2825-L2835

In my `IndirectParameter` prototype for FSDP, I changed `FlatParameter: nn.Parameter` to `FlatTensor: Tensor` since we no longer want to expose the `FlatParameter` to `nn.Module` methods like `named_parameters()`. In doing this, I still need to use the `expand_as()` trick since the `FlatTensor` is a leaf tensor. However, I find that the resulting tensor's `grad_fn` chain is now `AliasBackward0` -> `ExpandBackward0` -> `AccumulateGrad`; i.e. there is an additional `AliasBackward0`. I am not sure if this is expected. Either way, this provides some motivation for an API to register a post-gradient-computation hook (i.e. a hook that runs _after_ the gradient is computed) since this `expand_as()` trick is arguably hacky.

### Alternatives

The alternative is to continue using the `expand_as()` trick. When registering the post-backward hook for FSDP, we will need some logic to handle both the `FlatParameter` and `FlatTensor` case since we will need to preserve the existing code path in the near term. This is not a real problem.

### Additional context

In case this is relevant to the appearance of the `AliasBackward0`, the `FlatTensor`, which is just a `torch.Tensor`, is initialized like:
```
class FlatTensor(torch.Tensor):
    ...

@staticmethod
def flatten_params(
    params: List[torch.Tensor],
    requires_grad: bool,
) -> FlatTensor:
    with torch.no_grad():
        flat_tensors = [
            p.detach().reshape(-1) if isinstance(p, nn.Parameter) else p.reshape(-1)
            for p in params
        ]
        flat_tensor_data = torch.cat(flat_tensors, dim=0)
    flat_tensor = FlatTensor(flat_tensor_data)
    flat_tensor.requires_grad_(requires_grad)
    return flat_tensor
```

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @zhaojuanmao @mrshenli @rohan-varma

**Comment by albanD** [comment]
Duplicate of https://github.com/pytorch/pytorch/issues/76464

**Comment by awgu** [comment]
Ah sorry. I should really search the issues before posting.

**Comment by vadimkantorov** [comment]
a bit related not general flatten_parameters outside of FSDP: https://github.com/pytorch/pytorch/issues/38655#issuecomment-631383944

**Comment by awgu** [comment]
`register_post_accumulate_grad_hook()`!
